firedrake/mesh.py

```diff
@@ -1025,6 +1028,7 @@ def topological(self):
         return self._topology
 
     @utils.cached_property
+    @MeshGeometryMixin._ad_annotate_coordinates_function
```

Comment:
Does this play nicely with the cached_property? I noticed some comments above in the block file.

Reply:
Not really, this has caused slate's tests to fail because the caching was not done correctly and new coordinates were created for every call of the method. The reason is that `utils.cached_property` will not cache the method but the wrapper returned by the mixin's decorator. The name of the cached function will then be `wrapper`, that is why the name was hacked in the block file. This is a recurrent problem when dealing with composition of decorators.

Reply:
We should backport `cached_property` from functools in Py3.8 which handles this behaviour correctly. Can you try with OP2/PyOP2#573 ?

Reply:
Ok, I replied in the OP2 PR.